text/0000-multi-sig-term.md

```diff
@@ -0,0 +1,340 @@
+   ff  # break
+```
+
+## Multi-Sig term Size
+
+The term size can be calculated from the CDDL description above, by recursively
+summing up the sizes of the sub-terms. Every key hash requires `34` bytes, so
+the encoded keys generally make up the largest part of the size of the encoded
+terms. Currently we use indefinite length lists, but as the expected number of
+elements in each list is rather low, having an explicit length marker would
+allow using a single byte start marker with the length included.
+
+## Script Hashing
+
+The hash of a multi-sig term is computed on a tagged version of the binary
+representation. This allows for easy extension with other multi-signature
+schemes using different scripting languages in the future without the need to
+change the ledger rules.
+
+Therefore when hashing the example script for the required signature of Alice
+and Bob above, we prefix the CBOR encoded script with an additional tag `1` and
+vkey addresses with a tag `0`. Other tags will be used for other scripting
+languages.
+
+## Validation of Multi-Sig Terms
+
+The validation of a multi-signature script is done recursively over the
+structure of the term. The validator function also requires the set of
+key hashes `vkhs` for which the corresponding key has correctly signed the
+transaction.
+
+The validation works as follows
+
+- The term `RequireSignature kh` is evaluated as
+```validate (RequireSignature kh) vkhs = kh ∈ vhks```
+- The term `RequireAllOf [msigs]` is evaluated as
+```validate (RequireAllOf) [msigs] = all (λt. validate t vkhs) [msigs]```
+- The term `RequireAnyOf [msigs]` is evaluated as
+```validate (RequireAnyOf) [msigs] = any (λt. validate t vkhs) [msigs]```
+- The term `RequireMOf m [msigs]` is evaluated as
+```validate (RequireMOf m [msigs]) vkhs = m <= sum [if validate t vkhs then 1
+else 0 | t <- msigs]```
+
+# Drawbacks/Implications
+[drawbacks]: #drawbacksimplications
+[implications]: #drawbacksimplications
+
+Why should we *not* do this?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The approach to use a script-like DSL for multi-signature schemes allows for an
+easy extension point in two dimensions:
+
+ 1. additional features, e.g., hash locks and time locks, by extending the terms
+ 2. integration of other script languages
+
+The specification gives an outline how another script language can be integrated
+which is a first step in the direction of full smart contract support with
+Plutus. The required changes to the ledger rules and therefore also to the
+implementation will be trivial for multi-signature using simplified Plutus (no
+data and redeemer scripts) and minimal for full Plutus support.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The presented scheme does not resolve the problem of the communication that is
+necessary for parties to agree and collectively sign a transaction for a
+multi-signature scheme.
```

Comment:
And we've not yet addressed limiting the overall script size, or any relation to fees.

Reply:
I added this. Should we make this that "... any relation to fees apart from the additional size of the encoded script in bytes. As using multi-signature terms might incur significantly higher computational cost, this will require a specific cost-model." ?